### ISSUE TITLE
Community governance process is broken

### ISSUE THREAD

**Issue by Jaykul** [PowerShell/PowerShell#5862]
It's totally (**TOTALLY**) not ok to just suddenly make breaking changes 2 days before GA.

That's not how Open Source is done.

That's _certainly_ not how the community governance process works. According to [the governance docs](https://github.com/PowerShell/PowerShell/blob/master/docs/community/governance.md) and the [breaking changes contract](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md), it seems to me this should have actually required an RFC. But certainly any breaking change should have at least been made in a beta prior to being incorporated in a final release. 

I really can't say I care _that much_ about this particular alias, but I'm really frustrated and confused about the sudden about-face and the end-run on the process, bypassing any _community_ discussion of the change.

To summarize: 

1. In June, someone asked for a specific alias (which has been in PowerShell for a decade) to be removed because it hides an executable they like to use. 
2. The request was rejected and closed.
3. 6 months later, someone comments on the issue
4. Totally off topic posts on the issue for a few days
5. TWO DAYS before final GA, @SteveL-MSFT suddenly un-closes the issue and removes the alias in a PR which flies through all code reviews and is merged in just three hours. Nobody even notices.
6. Now it shows up in the release notes as a breaking change, because yeah, you just broke every Windows PowerShell **or** PowerShell Core script that's _ever_ been written that that calls `sc` ....

Why is the `sc` alias being removed?

Hint: don't say that it's because it shadows a native command: `fc` and `where` and `sort` and `write` also shadow commands but were not removed. Additionally, these aliases have been shadowing native commands since they were introduced, and they were not removed back when the last set of alias purges were implemented .... 

For what it's worth, I *do* use that alias all the time, and now I have to patch my profiles to create yet another function that yells at me, so I can learn not to type `sc` -- to avoid screwing up when working on someone else's dev box or on a server. I already have a significant amount of pain in PowerShell Core with aliases being defined on some OS's but not others, and this is just yet another papercut.

At this point I'm convinced that you should never have shipped built-in aliases at all. It would have been easier for me to have never learned them.

**Comment by SteveL-MSFT** [comment]
I think you have a fair criticism, but it's unfair to state that I made this call by myself when I clearly wrote in the PR that this was discussed with the @PowerShell/powershell-committee members and we agreed this was a change we should take.

It was unfortunate timing that this happened very late, but one of the decision factors was that if we made the change now and feedback is that we got it wrong.  We can put it back for 6.1.0.  But if we left it, after GA the bar for breaking changes is significantly higher and it would effectively be there forever.

I think the principal of least astonishment applies and in this case using `sc start winrm` appears to have worked as no error is reported, but instead created a file called `start` containing `winrm` because it was aliases to `Set-Content`.

It's unfortunate that the other examples you bring up were not discussed along with `sc`.  However, in those cases the usage would have returned an error making it clear that you used the alias and not the native tool which is different from the `sc.exe` case.

**Comment by iSazonov** [comment]
In fact, the team does not have as much resources to provide internal development and GITHUB and support the full process including RFC. It's a fact. The current short process is quite good - a discussion in Issue or PR and the PowerShell Committee's conclusion. While PowerShell Core is just a small subset of Windows PowerShell and is still a toy we are doomed to make a lot of breaking changes very quickly. I even suggested to explicitly indicate this in our documentation.

**Comment by swngdnz** [comment]
sc should not have been un-aliased. I just checked my scripts, and I've got 27 uses of it, in scripts written over the last 10 years.  More breakage when trying to move to PS6. Every single time I wanted to use sc.exe I specified sc.exe. **Any** time I want to use a native binary I specify _exe_.

**Comment by markekraus** [comment]
@swngdnz I have add issue #5866 to request `sc` be re-added. Please comment/vote there

**Comment by Jaykul** [comment]
@iSazonov It's ok to say "we don't have the resources to do an RFC for everything" ...

It's **not** ok to create a breaking changes contract and then ignore it.

It's not ok, _in any product_ that has a beta cycle, to make **breaking changes** _after_ the beta cycle. That's **just _plain_ wrong**. If this team is going to do that, then PowerShell will become an undependable, unusable product.

@SteveL-MSFT I acknowledge it went through code review (very quickly, and at the last minute). If you want to throw someone else under the bus, feel free :wink:. It's not my intention to lambast you as if you were saving this secret change until the last minute and sneaking it in to deliberately make things more painful. If you like, my problem is that _the whole team_ allowed a rather egregious violation of process.  This is a bucket 1, public contract, breaking change. It was done **despite** prior community discussions which had resulted in the change being rejected, and it was done without any pre-release.  

I disagree with the argument that it's easier to put it back later than to take it out later, but even if I agreed, that's a *terrible* reason to make a breaking change that late in the development cycle.

**Comment by Jaykul** [comment]
Just for the record, and to vent, regarding this particular change:

When someone suggests that the "principle of least astonishment" argues for _removal_ of an existing alias, it feels a bit like an insult. That person doesn't seem to care about me, nor about my friends and co-workers. They are ignoring the user base of people who've been using, learning, and promoting PowerShell for years, and ignoring our code-base of existing code.  There are a lot of us, and a lot of  our code. **That's a lot of astonishment.**

Additionally, in this particular case, if "we" accidentally change a service because we called `sc.exe` inadvertently, that's going to be a lot more painful than if a new user accidentally creates a file called "config" or "start" because they called `Set-Content` without realizing it...

When someone says "you shouldn't use aliases in your scripts" it feels patronizing and misguided. It's the wrong argument at the wrong time. The *only* reason we have said in the past that you shouldn't use aliases in scripts is for _readability_. It was never about breaking changes, because the built-in aliases have always been (until Core) dependable:

1. They are always defined in a clean environment, even before any modules are imported
2. The backward compatibility commitment from Microsoft meant they _would always_ be there --as they have been, for over a decade.

In fact, because they are hard-coded and ReadOnly, it has always been **harder** for a user to _change_ a built-in alias than to overwrite a command with a function from a module or with an alias! Built-in aliases are therefore as safe and dependable as any PowerShell code can be, and so the use of them was down to readability. 

I've made this argument frequently: _in many cases, aliases are more readable than the full command_, especially years ago when Windows Powershell was new, and we were a group of people familiar with the old cmd and bash utilities. An alias like `dir` or `ls` was so much clearer than the new `Get-ChildItem` command. Not just easier to type, but readable to a much broader range of people.

Anyway. Having vented all of that, we don't need to debate this particular alias. What's done is done. It's shipped. It's broken, and it can't be unbroken.

## The question is: 

Does the PowerShell Team intend to following the guidelines and live up to the breaking changes commitment, or is that just, as Ilya says, too much work?

If you're not going to do it, it would be prudent to rewrite your documents to say what you mean -- better to anger a people over a documentation change than over broken promises.

**Comment by joeyaiello** [comment]
Thanks for being so frank, everyone. This is clearly a controversial issue, and I appreciate everyone for providing your input and being level-headed. 

I wholeheartedly agree that we screwed up here from a governance and timing perspective. While @iSazonov is right that we don't require RFCs for smaller changes like this (smaller in the sense that the change doesn't require a design, not in the sense of the *scope* of the impact), it doesn't excuse the fact that we violated the *spirit* of our governance model. And I certainly agree that we should update our governance documentation to codify the reality of the RFC vs. small change process. 

@Jaykul is also right in that we didn't plan on sneaking this in, and that it came up at the 11th hour. I'm very sorry that I missed the initial conversation on `sc` from way back when (or maybe I just don't remember), but I promise you that this was a case of "oh, crap, we missed something" rather than a sneaky move.

All that being said, we did have a long discussion on the implications of this, and even with the community input, I still would've personally made the same decision. While controversial, [our stance on aliases has been pretty clear over the last few years](https://github.com/PowerShell/PSScriptAnalyzer/issues/214): namely, that they're almost purely for interactive usage, that they shouldn't and can't be used to provide truly deterministic behavior in scripts/modules shared outside of your personal usage, and that you should add aliases as appropriate to your profiles or within module scope if you want them. 

For example, @JamesWTruher uses all of the POSIX-y aliases like `ls` and `cp` on his Mac, and he put them back in his profile the day we took them out. We *encourage* this. And as the [WindowsPowerShellCompatibilityPack](https://github.com/PowerShell/WindowsPowerShellCompatibilityPack) gets developed, there'll be mechanisms to make it easier to put them back where you want it.

And while I know that your inclination is to say "well just take them all out!" (I'm looking at #5870), aliases are *way* more egregious when they collide with native executables. It's how we ended up with the curl/wget problem and it's why we removed the POSIX-y aliases on macOS/Linux. I understand that many of you have worked around it, built your muscle memory, and understand what's going on, but @SteveL-MSFT is right in that the lack of an error message from Set-Content makes this really hard to figure out.

So in summary: I'm very sorry that we about removing `sc` like this, and I completely understand the frustration coming from all of you, as this issue is definitely not cut-and-dry. We had to make hard decisions on breaking some stuff, and weighing that against back compat and potential adoption. I hope you also understand, though, that we truly didn't make this one lightly (even if that's how it appeared).

**Comment by Jaykul** [comment]
For what it's worth, @SteveL-MSFT -- if you look at [that thread](https://github.com/PowerShell/PSScriptAnalyzer/issues/214) closely, you'll see that the reason that thread is there is that someone wrote a PSScriptAnalyzer rule saying that aliases are bad -- and the community feedback (of which I was, obviously, a loud voice) was that this is not true, and that many aliases improve readability of code.  The issue you linked to is actually the response to our frustration: an attempt to keep the rule but create a white-list of aliases that we think are better than writing the full command name. You'll also notice the _**complete** lack of action_ on that THREE YEAR OLD issue -- which is why my default mode when it comes to aliases is frustration.

My argument at this point is that if you're not committed to _stop changing them_, you should _completely remove them_, instead of changing them randomly when someone points out a collision with something they care about. Your commitment to remove everything that _anyone notices_ collides is breaking your commitment to keep my old code working, and it's very frustrating.

**Comment by SteveL-MSFT** [comment]
@Jaykul I hadn't seen that thread previously as PSScriptAnalyzer was owned by a different engineering manager so I wasn't aware of that discussion.  Recently I did become owner of PSScriptAnalyzer, so over the next few months at least you'll see more activity from my team in that repo and such discussions can get more traction more quickly.

**Comment by rkeithhill** [comment]
FWIW the ability to whitelist aliases was added to PSSA:
```
Rules = @{
    # Do not flag 'cd' alias.
    PSAvoidUsingCmdletAliases = @{Whitelist = @('cd')}
}
```

**Comment by Jaykul** [comment]
I didn't know that. "Someone" should close that issue ;-)

**Comment by bwright86** [comment]
I feel @joeyaiello is right, in that aliases are only for interactive use. This is practically hammered into beginner’s heads when they seek advice online. Any scripts that depend on aliases should not be considered a breaking change. In the Shell, I see the benefit to shorten a multi liner command with aliases. But in a script, it is better to be verbose and use the full Cmdlet name, than to save characters.

I think this example should be a wake up call to that effect.

**Comment by iSazonov** [comment]
@bwright86 Feel free to discuss the RFC https://github.com/PowerShell/PowerShell-RFC/pull/129
